compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp

```diff
@@ -50,59 +51,52 @@ static FailureOr<Operation *> getRootOp(ArrayRef<Operation *> computeOps) {
-/// 1}.
-static LogicalResult lowerToUnitWorkgroupCount(
-    IREE::Flow::DispatchWorkgroupCountFromDagRootOp workgroupCountOp) {
-  OpBuilder builder(workgroupCountOp.getContext());
-  builder.setInsertionPoint(workgroupCountOp);
-  Value one =
-      builder.create<arith::ConstantIndexOp>(workgroupCountOp->getLoc(), 1);
-  SmallVector<Value> replacements(workgroupCountOp->getNumResults(), one);
-  workgroupCountOp->replaceAllUsesWith(replacements);
-  workgroupCountOp.erase();
-  return success();
-}
-
-/// Method to lower the `flow.dispatch.workgroup_count_from_dag_root` op into
-/// the actual computation that returns the number of workgroups.
-static LogicalResult lowerDispatchWorkgroupCountFromDagRootOp(
-    IREE::Flow::DispatchWorkgroupCountFromDagRootOp workgroupCountOp,
+/// Method to return the configuration to use for first-level tile and
+/// distribute. Returns the
+/// - tileSizes to use
+/// - intechange
+/// - loops to be partitioned (the tile sizes for the non-partitioned loop are
+///   set to 0)
+/// - static loop ranges - this is meant to be an optimization hint. It recovers
+///   the static values that the workload of the dispatch corresponds to.
+// TODO: Remove the use of static loop ranges. This is used to set the number of
+// workgroups to a static value. Ideally this should not be done and the static
+// and dyamic cases are handled the same way. When the tile+distribute moves
+// away from using `scf.for` to using a construct that better captured
+// distribution (like `scf.foreach_thread`) this information can be dropped.
+static LogicalResult getTileAndDistributeConfig(
     ArrayRef<Operation *> computeOps, SmallVectorImpl<int64_t> &tileSizes,
-    SmallVector<int64_t> &interchange) {
-  auto workloadValues = workgroupCountOp.operands();
-
+    SmallVectorImpl<int64_t> &staticLoopRanges,
+    SmallVectorImpl<int64_t> &interchange,
+    SmallVector<unsigned> &partitionableLoops) {
   // Find the lowering configuration of the root operation.
   FailureOr<Operation *> rootOp = getRootOp(computeOps);
   if (failed(rootOp)) {
-    return lowerToUnitWorkgroupCount(workgroupCountOp);
+    // Just return. All the in-out vectors are empty that should default
+    // the number of workgroups to {1, 1, 1}
+    return success();
+    ;
   }
 
   auto partitionableLoopInterface =
       dyn_cast<PartitionableLoopsInterface>(*rootOp);
   if (!partitionableLoopInterface) {
-    return lowerToUnitWorkgroupCount(workgroupCountOp);
+    // Just return. All the in-out vectors are empty that should default
+    // the number of workgroups to {1, 1, 1}
+    return success();
   }
 
-  SmallVector<unsigned> partitionableLoops =
+  partitionableLoops =
       partitionableLoopInterface.getPartitionableLoops(kNumMaxParallelDims);
+  // For now assert that number of partitionable loops are less than the
+  // supported max.
+  // TODO(ravishankarm): Relax this restriction.
+  if (partitionableLoops.size() > kNumMaxParallelDims) {
```

Comment:
I don't think this case is possible because `getPartitionableLoops()` trims the excess.